### ISSUE TITLE
false positive needless_lifetimes when self is &Arc<Self>

### ISSUE THREAD

**Issue by antialize** [rust-lang/rust-clippy#7296]
Lint name: needless_lifetimes

I tried this code:

```rust
struct Stuff<'a>(std::marker::PhantomData<&'a ()>);
struct Thing {}
impl Thing {
    fn mut_stuff<'a>(self: &'a std::sync::Arc<Self>) -> Stuff<'a> {todo!()}
}
```

Clippy complains that <'a> is a needless life time, but it appears to not be the case. None of the following alternatives compile:
```rust
fn mut_stuff(self: &std::sync::Arc<Self>) -> Stuff<'_> {todo!()}
fn mut_stuff(self: &'_ std::sync::Arc<Self>) -> Stuff<'_> {todo!()}
fn mut_stuff(self: & std::sync::Arc<Self>) -> Stuff {todo!()}
```

# Meta

Build using the Nightly version: 1.54.0-nightly

(2021-05-30 c1e8f3a5857c3c8d095f)

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=12f0ac7dd5b7f413dffe26a765343475
